Telemetry / Upgrade Responder Integration

irq0 opened the pull request:
Summary:
- Start background thread that periodically POSTs to the s3gw upgrade responder URL
- Parses some of the response to get the list of available versions
- Add status page showing service status, available versions and the request data


Issue: https://github.com/aquarist-labs/s3gw/issues/423

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

Commit:
rgw/s3gw: Add telemetry / upgrade responder service

Add S3GWTelemetry "service" - an object sending periodic requests to
our upgrade responder instance and keeping the response for later
consumption.

Add status page '/telemetry' exporting upgrade responder request,
S3GWTelemetry status information and the available versions.

Example request (devel build):
```
{
    "appVersion": "18.0.0-3118-ge21d2ab4b05",
    "extraInfo": {
        "ceph_version": "ceph version 18.0.0-3118-ge21d2ab4b05 (e21d2ab4b0581a3aaeacaab2ef597ab41356b243) reef (dev)",
        "ceph_version_short": "18.0.0-3118-ge21d2ab4b05",
        "ceph_release": "reef",
        "os": "Linux",
        "kernel_version": "6.1.7-1-default",
        "kernel_description": "#1 SMP PREEMPT_DYNAMIC Wed Jan 18 11:12:34 UTC 2023 (872045c)",
        "arch": "x86_64",
        "mem_total_kb": "32486064",
        "mem_swap_kb": "32485372",
        "cpu": "Intel(R) Core(TM) i9-10885H CPU @ 2.40GHz",
        "cgroup_mem_limit": "0",
        "sfs_avail_kb": 43152412,
        "sfs_total_kb": 300933564,
        "sfs_used_kb": 257781152,
        "sfs_fs": "xfs"
    }
}
```

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
rgw/s3gw: Add s3gw telemetry unit test

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
ci: Include s3gw matching unittests to sfs/build-radosgw.sh

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

irq0 commented:
back to draft to incorporate feedback from https://github.com/aquarist-labs/s3gw/issues/202

votdev commented on src/rgw/rgw_status_page_telemetry.h:
```diff
@@ -0,0 +1,38 @@
+ * Ceph - scalable distributed file system
+ * SFS SAL implementation
+ *
+ * Copyright (C) 2022 SUSE LLC
```
```suggestion
 * Copyright (C) 2023 SUSE LLC
```

votdev commented on src/rgw/rgw_status_page_telemetry.cc:
```diff
@@ -0,0 +1,108 @@
+ * Ceph - scalable distributed file system
+ * SFS SAL implementation
+ *
+ * Copyright (C) 2022 SUSE LLC
```
```suggestion
 * Copyright (C) 2023 SUSE LLC
```

Commit:
ci: add new files to pre-commit files list

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

jecluis reviewed (dismissed):
would love to do some manual testing before merging, but the code looks good

Commit:
rgw/s3gw: Add telemetry / upgrade responder service

Add S3GWTelemetry "service" - an object sending periodic requests to
our upgrade responder instance and keeping the response for later
consumption.

Add status page '/telemetry' exporting upgrade responder request,
S3GWTelemetry status information and the available versions.

Example request (devel build):
```
{
    "appVersion": "18.0.0-3118-ge21d2ab4b05",
    "extraInfo": {
        "ceph_version": "ceph version 18.0.0-3118-ge21d2ab4b05 (e21d2ab4b0581a3aaeacaab2ef597ab41356b243) reef (dev)",
        "ceph_version_short": "18.0.0-3118-ge21d2ab4b05",
        "ceph_release": "reef",
        "os": "Linux",
        "kernel_version": "6.1.7-1-default",
        "kernel_description": "#1 SMP PREEMPT_DYNAMIC Wed Jan 18 11:12:34 UTC 2023 (872045c)",
        "arch": "x86_64",
        "mem_total_kb": "32486064",
        "mem_swap_kb": "32485372",
        "cpu": "Intel(R) Core(TM) i9-10885H CPU @ 2.40GHz",
        "cgroup_mem_limit": "0",
        "sfs_avail_kb": 43152412,
        "sfs_total_kb": 300933564,
        "sfs_used_kb": 257781152,
        "sfs_fs": "xfs"
    }
}
```

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

irq0 commented:
Last push fixes the previous formatting error
